ocaml/libs/tracing/tracing.ml

```diff
@@ -574,38 +574,90 @@ module Export = struct
-          in
-          Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file) 0o700 ;
-          Xapi_stdext_unix.Unixext.write_string_to_file file span_json ;
-          Ok ""
-        with e -> Error e
+      let set_max_file_size size = max_file_size := size
+
+      let file_name = ref None
+
+      let file_stream = ref None
+
+      let lock = Mutex.create ()
+
+      let new_file_name () =
+        let date = Ptime_clock.now () |> Ptime.to_rfc3339 ~frac_s:6 in
+        let ( // ) = Filename.concat in
+        let name =
+          !trace_log_dir
+          // String.concat "-" [get_service_name (); !host_id; date]
+          ^ ".json"
+        in
+        file_name := Some name ;
+        name
+
+      let open_stream file_name =
+        Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file_name) 0o700 ;
+        file_stream :=
+          Some (Unix.openfile file_name [O_WRONLY; O_CREAT; O_APPEND] 0o700)
+
+      let close_stream () =
+        Option.iter (fun stream -> Unix.close stream) !file_stream ;
+        file_stream := None
+
+      let check () =
+        Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+            match (!file_stream, !file_name) with
+            | Some file_stream, _
+              when (Unix.fstat file_stream).st_size >= !max_file_size ->
+                close_stream () ;
+                new_file_name () |> open_stream
+            | None, None ->
+                new_file_name () |> open_stream
+            | None, Some file_name ->
+                open_stream file_name
+            | _ ->
+                (* We have an open fd which points to a file under the limit*)
+                ()
+        )
+
+      let write str =
+        Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+            Option.iter
+              (fun stream ->
+                let content = Bytes.of_string (str ^ "\n") in
+                ignore @@ Unix.write stream content 0 (Bytes.length content)
+              )
+              !file_stream
+        )
+
+      let export ~span_json =
+        try check () ; write span_json ; Ok () with e -> Error e
+
+      let with_stream f =
+        f export ;
+        Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+            close_stream ()
+        )
     end
 
     module Http = struct
       module Request = Cohttp.Request.Make (Cohttp_posix_io.Buffered_IO)
       module Response = Cohttp.Response.Make (Cohttp_posix_io.Buffered_IO)
 
-      let export ~span_json ~url : (string, exn) result =
-        try
-          let body = span_json in
-          let headers =
-            Cohttp.Header.of_list
-              [
-                ("accepts", "application/json")
-              ; ("content-type", "application/json")
-              ; ("content-length", string_of_int (String.length body))
-              ]
-          in
-          Open_uri.with_open_uri url (fun fd ->
+      let export ~url ~span_json =
+        let body = span_json in
```

Comment:
we probably want to keep the try statement at the beginning of the function, so that all the call to libraries below are protected by it.